torchio/data/io.py

```diff
@@ -1,88 +1,84 @@
-            f' but they are {suffixes}')
-    return path
-
-
-def read_image(path: TypePath) -> Tuple[torch.Tensor, np.ndarray]:
-    path = parse_path(path)
-    suffixes = path.suffixes
-    if '.nii' in suffixes:
-        read = _read_nifti
-    elif '.nrrd' in suffixes:
-        read = _read_nrrd
+def read_image(
+        path: TypePath,
+        itk_first: bool = False,
+        ) -> Tuple[torch.Tensor, np.ndarray]:
+    if itk_first:
+        try:
+            result = _read_sitk(path)
+        except RuntimeError:  # try with NiBabel
+            result = _read_nibabel(path)
     else:
-        raise NotImplementedError(
-            f'Reading not implemented for this format: "{path}"')
-    return read(path)
+        try:
+            result = _read_nibabel(path)
+        except nib.loadsave.ImageFileError:  # try with ITK
+            result = _read_sitk(path)
+    return result
 
 
-def _read_nifti(path: TypePath) -> Tuple[torch.Tensor, np.ndarray]:
+def _read_nibabel(path: TypePath) -> Tuple[torch.Tensor, np.ndarray]:
     nii = nib.load(str(path), mmap=False)
-    ndims = len(nii.shape)
-    assert ndims == 3
     data = nii.get_fdata(dtype=np.float32)
     tensor = torch.from_numpy(data)
     affine = nii.affine
     return tensor, affine
 
 
-def _read_nrrd(path: TypePath) -> Tuple[torch.Tensor, np.ndarray]:
-    data, header = nrrd.read(path)
-    data = data.astype(np.float32)
+def _read_sitk(path: TypePath) -> Tuple[torch.Tensor, np.ndarray]:
+    image = sitk.ReadImage(str(path))
+    data, affine = sitk_to_nib(image)
+    if data.dtype != np.float32:
+        data = data.astype(np.float32)
     tensor = torch.from_numpy(data)
-    affine = np.eye(4)
-    affine[:3, :3] = header['space directions'].T
-    affine[:3, 3] = header['space origin']
-    lps_to_ras = np.diag((-1, -1, 1, 1))
-    affine = np.dot(lps_to_ras, affine)
     return tensor, affine
 
 
 def write_image(
         tensor: torch.Tensor,
         affine: TypeData,
         path: TypePath,
+        itk_first: bool = False,
         ) -> None:
-    path = Path(path)
-    suffixes = path.suffixes
-    if '.nii' in suffixes:
-        write = _write_nifti
-    elif '.nrrd' in suffixes:
-        write = _write_nrrd
+    if itk_first:
+        try:
+            _write_sitk(tensor, affine, path)
+        except RuntimeError:  # try with NiBabel
+            _write_nibabel(tensor, affine, path)
     else:
-        raise NotImplementedError(
-            f'Writing not implemented for this format: "{path}"')
-    write(tensor, affine, path)
+        try:
+            _write_nibabel(tensor, affine, path)
+        except nib.loadsave.ImageFileError:  # try with ITK
+            _write_sitk(tensor, affine, path)
 
 
-def _write_nifti(
+def _write_nibabel(
         tensor: torch.Tensor,
         affine: TypeData,
         path: TypePath,
         ) -> None:
+    """
+    Expects a path with .nii or .nii.gz extension, and writes a NIfTI-1 file
+    with qform only
```

Comment:
Since Nibabel can theoretically write out other file formats (though often conversion isn't implemented), should be "Expects a path with an extension  such as '.nii' or 'nii.gz' to which `nibabel.save` supports writing Nifti1 files" or similar?

Reply:
Well, the issue is that my code is now written to create a nib.Nifti1Image object. Do you know if there is a way to let NiBabel choose the class depending on the extension?

Reply:
Yeah, I'm not sure.  You could try to do something with `nib.imageclasses.class_map` but some of the extensions overlap, ".mnc" maps to an ancient MINC1 class, etc ... maybe it's fine to keep as-is for now and just tweak the docstring slightly?

Reply:
Sounds a bit overkill, especially for those formats that are, I believe, mostly quite old. Do you use MINC? If people ask for it, I'll consider working on this. That `class_map` seems like a good resource if ever needed.

Reply:
I see what you mean by "to which nibabel.save supports writing Nifti1 files". One could also use '.hdr' or '.img.gz' (although why on earth?!).

Reply:
Yeah, seems SimpleITK is ahead in terms of format support.  I do use MINC but don't really need special support -- I can just convert first or use SITK.